core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java

```diff
@@ -926,30 +1011,57 @@ public Command workerHeartbeat(long workerId, Map<String, Long> capacityBytesOnT
       return Command.newBuilder().setCommandType(CommandType.Register).build();
     }
 
-    synchronized (worker) {
-      // Technically, 'worker' should be confirmed to still be in the data structure. Lost worker
-      // detection can remove it. However, we are intentionally ignoring this race, since the worker
-      // will just re-register regardless.
-      processWorkerRemovedBlocks(worker, removedBlockIds);
-      processWorkerAddedBlocks(worker, addedBlocks);
-      processWorkerMetrics(worker.getWorkerAddress().getHost(), metrics);
+    // Update the TS before the heartbeat so even if the worker heartbeat processing
+    // is time-consuming or triggers GC, the worker does not get marked as lost
+    // by the LostWorkerDetectionHeartbeatExecutor.
+    worker.updateLastUpdatedTimeMs();
+
+    // The address is final, no need for locking
+    processWorkerMetrics(worker.getWorkerAddress().getHost(), metrics);
 
+    worker.getUsageLock().writeLock().lock();
+    worker.getBlockListLock().writeLock().lock();
+    Command workerCommand = null;
+    try {
       worker.addLostStorage(lostStorage);
 
       if (capacityBytesOnTiers != null) {
         worker.updateCapacityBytes(capacityBytesOnTiers);
       }
       worker.updateUsedBytes(usedBytesOnTiers);
-      worker.updateLastUpdatedTimeMs();
 
+      // Technically, 'worker' should be confirmed to still be in the data structure. Lost worker
+      // detection can remove it. However, we are intentionally ignoring this race, since the worker
+      // will just re-register regardless.
+
+      // This changes the block list
+      processWorkerRemovedBlocks(worker, removedBlockIds);
+      processWorkerAddedBlocks(worker, addedBlocks);
+      // TODO(jiacheng): If the worker registers with a after the block is freed,
+      //  the copy will not be removed from this worker.
```

Comment:
will this be addressed in this PR?

Reply:
This will not be addressed in this PR.

On 2nd thought, I'm not sure if this behavior should be marked as "expected". 

The sequence is like this:
1. `removeBlocks(blockList, false)` is called where the boolean field means only freeing the blocks but not deleting metadata
2. `removeBlocks` gets the current block locations from the `BlockStore`. This step holds the on block lock. The block lock is now released. The locations are stored in variable `workerIds`.
3. Worker comes in with a `register`. This gets the block lock and updates the location in `BlockStore`.
4. `removeBlocks` loops `workerIds` and adds this blockId to the corresponding `MasterWorkerInfo`. This step holds the `MasterWorkerInfo` locks.
5. `removeBlocks` completes, the copy on the worker is never marked as to-be-removed.

Step 3 and 4 are concurrent and orthogonal.

Reply:
To clarify more, when the `registerWorker` is concurrent with `removeBlocks(blockList, true)` where `true` indicates the block metadata are removed, the block on the worker WILL be removed correctly. The behavior above only happens when the operations is a `free`.

Reply:
I feel this behavior is actually okay. The concurrent `registerWorker` is not very different from a worker that registers way after the `free`. The block can stay. We don't have a `free-and-no-cache` semantic.

@ggezer What do you think?

Reply:
I see. It looks like we could avoid this if we acquire the block lock outside of process* methods. (Actually it might not be a bad idea to merge them into a single method as register and heartbeat have the same signature.)

However it's too much effort for something that is inevitable under different scenarios as you pointed out.
Please update this TODO with a summary of what you mentioned, or it might not make sense later.

Reply:
Added TODO to `removeBlocks()`